var/spack/repos/builtin/packages/sgpp/package.py

```diff
@@ -0,0 +1,167 @@
+
+    # Patch that ensures libraries will actually
+    # be copied into prefix/lib upon installation
+    # (otherwise it would be prefix/sgpp/lib)
+    patch('directory.patch', when='@:3.2.0')
+    # Fix faulty setup.py in 3.2.0
+    patch('fix-setup-py.patch', when='@3.2.0')
+    # Backport opencl fix
+    patch('ocl.patch', when='@:3.2.0+opencl')
+
+    variant('simd',
+            default='sse3',
+            values=('sse3', 'sse42', 'avx', 'fma4', 'avx2', 'avx512'),
+            description='Specifies the SIMD extension to be used')
+    variant('python', default=True,
+            description='Provide Python bindings for SGpp')
+    variant('java', default=False,
+            description='Provide Java bindings for SGpp')
+    # export CLASSPATH=$(spack location --install-dir sgpp)/lib/jsgpp.jar
+    # After that one can compile and run simple java examples with sgpp
+    variant('optimization', default=True,
+            description='Builds the optimization module of SGpp')
+    variant('pde', default=True,
+            description='Builds the datadriven module of SGpp')
+    variant('quadrature', default=True,
+            description='Builds the datadriven module of SGpp')
+    variant('datadriven', default=False,
+            description='Builds the datadriven module of SGpp')
+    variant('misc', default=False,
+            description='Builds the misc module of SGpp')
+    variant('combigrid', default=False,
+            description='Builds the combigrid module of SGpp')
+    variant('solver', default=True,
+            description='Builds the solver module of SGpp')
+    variant('opencl', default=False,
+            description='Enables support for OpenCL accelerated operations')
+    variant('mpi', default=False,
+            description='Enables support for MPI-distributed operations')
+
+    # Mandatory dependencies
+    depends_on('scons@2.5.1', when='@:3.1.0', type=('build'))
+    depends_on('scons@3:', when='@3.2.0:', type=('build'))
+    depends_on('zlib', type=('build', 'run'))
+    # Python dependencies
+    extends('python', when='+python')
+    # Python 3 support was added in version 3.2.0
+    depends_on('python@2.7:2.8', when='@:3.1.0+python', type=('build', 'run'))
+    depends_on('python@3:', when='@3.2.0:+python', type=('build', 'run'))
+    depends_on('swig@3:', when='+python', type=('build'))
+    # Java dependencies
+    depends_on('swig@3:', when='+java', type=('build'))
+    depends_on('openjdk', when='+java', type=('build', 'run'))
+    # Python libraries (version depends on whether we use Python 2 or 3)
+    depends_on('py-numpy@:1.16', when='@:3.1.0+python', type=('build', 'run'))
+    depends_on('py-numpy@1.17:', when='@3.2.0:+python', type=('build', 'run'))
+    depends_on('py-scipy@:1.2.3', when='@:3.1.0+python', type=('build', 'run'))
+    depends_on('py-scipy@1.3.0:', when='@3.2.0:+python', type=('build', 'run'))
```

Comment:
I thought this would work
```python
    depends_on('py-numpy@1.16.0:', when='+python', type=('build', 'run'))
    depends_on('py-scipy@1.2.3:', when='+python', type=('build', 'run'))
```
as the minimum listed versions work for both Python 2 and 3. However, if I try this with the Python 2 version of sgpp ```spack spec sgpp@3.1.0```, it still tries to use the newest numpy (1.18.2) instead of one that still supports Python 2 yielding the error ```py-numpy requires python version 3.5:, but spec asked for 2.7.16```

Fixing the version to an older one like this:
```python
    depends_on('py-numpy@1.16.0', when='+python', type=('build', 'run'))
    depends_on('py-scipy@1.2.3', when='+python', type=('build', 'run'))
```
would work for both Python 3 and Python 2. 

That being said: Unless I am missing any downsides to it, I would prefer to simply leave it as it is now! This way we will always get the newest numpy version for the more current SGpp versions and only use the older numpy versions for the old, python2-versions of SGpp.

Reply:
Yeah, that's a bug in Spack's concretizer that will be fixed when the new concretizer is released. I'm fine with leaving it as is until that bug has been fixed.